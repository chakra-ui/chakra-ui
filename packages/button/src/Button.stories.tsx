import { AddIcon } from "@chakra-ui/icons"
import { Stack } from "@chakra-ui/layout"
import * as React from "react"
import { Button } from "."

export default {
  title: "Button",
}

export const Variants = () => (
  <Stack direction="row" spacing="24px">
    <Button colorScheme="teal" variant="solid">
      Button
    </Button>
    <Button colorScheme="teal" variant="outline">
      Button
    </Button>
    <Button colorScheme="teal" variant="ghost">
      Button
    </Button>
    <Button colorScheme="teal" variant="link">
      Button
    </Button>
  </Stack>
)

export const Sizes = () => (
  <Stack direction="row">
    <Button colorScheme="blue" size="xs">
      Button
    </Button>
    <Button colorScheme="blue" size="sm">
      Button
    </Button>
    <Button colorScheme="blue" size="md">
      Button
    </Button>
    <Button colorScheme="blue" size="lg">
      Button
    </Button>
  </Stack>
)

export const WithCustomIcon = () => <Button leftIcon={AddIcon}>Call Us</Button>

export const WithLoading = () => (
  <Button colorScheme="pink" isLoading loadingText="Loading...">
    Pink Button
  </Button>
)

// export const WithButtonGroup = () => (
//   <ButtonGroup variant="solid">
//     <Button colorScheme="green">Save</Button>
//     <Button variant="outline">Cancel</Button>
//   </ButtonGroup>
// )

// export const iconButton = () => (
//   <ButtonGroup>
//     <IconButton fontSize="24px" aria-label="Add to friends" icon={AddIcon} />
//     <IconButton aria-label="Call us now" isRound icon={PhoneIcon} />
//   </ButtonGroup>
// )
