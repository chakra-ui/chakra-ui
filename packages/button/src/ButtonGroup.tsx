import { SystemProps, chakra, PropsOf, ThemingProps } from "@chakra-ui/system"
import { getValidChildren } from "@chakra-ui/utils"
import * as React from "react"
import { ButtonProps } from "./Button"

export type ButtonGroupOptions = Omit<ThemingProps, "orientation"> & {
  /**
   * If `true`, the borderRadius of buttons will be altered
   * to look flushed together
   */
  isAttached?: boolean
  /**
   * Spacing between each button
   */
  spacing?: SystemProps["marginRight"]
  /**
   * The buttons to group
   */
  children?: React.ReactNode
}

export type ButtonGroupProps = PropsOf<typeof chakra.div> & ButtonGroupOptions

export const ButtonGroup = ({
  size,
  colorScheme,
  variant,
  isAttached,
  spacing = 2,
  children,
  ...rest
}: ButtonGroupProps) => {
  const validChildren = getValidChildren(children)

  const clones = validChildren.map((child, index) => {
    const isFirst = index === 0
    const isLast = index === validChildren.length - 1

    return React.cloneElement(child as React.ReactElement<ButtonProps>, {
      size,
      colorScheme: child.props.colorScheme || colorScheme,
      variant: child.props.variant || variant,
      _focus: { boxShadow: "outline", zIndex: 1 },
      ...(!isLast && !isAttached && { marginRight: spacing }),
      ...(isFirst && isAttached && { borderRightRadius: 0 }),
      ...(isLast && isAttached && { borderLeftRadius: 0 }),
      ...(!isLast && isAttached && { borderRightRadius: 0 }),
      ...(!isFirst && !isLast && isAttached && { borderRadius: 0 }),
    })
  })

  return <chakra.div display="inline-block" {...rest} children={clones} />
}
