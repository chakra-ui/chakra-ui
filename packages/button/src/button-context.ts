import { createContext } from "@chakra-ui/react-utils"
import { ThemingProps } from "@chakra-ui/system"

export interface ButtonGroupContext extends ThemingProps<"Button"> {
  isDisabled?: boolean
}

export const [ButtonGroupProvider, useButtonGroup] =
  createContext<ButtonGroupContext>({
    strict: false,
    name: "ButtonGroupContext",
  })
