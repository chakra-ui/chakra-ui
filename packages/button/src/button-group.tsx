import {
  chakra,
  forwardRef,
  SystemProps,
  SystemStyleObject,
  ThemingProps,
  HTMLChakraProps,
} from "@chakra-ui/system"
import { cx, __DEV__ } from "@chakra-ui/utils"
import { createContext } from "@chakra-ui/react-utils"
import * as React from "react"

export interface ButtonGroupProps
  extends HTMLChakraProps<"div">,
    ThemingProps<"Button"> {
  /**
   * If `true`, the borderRadius of button that are direct children will be altered
   * to look flushed together
   */
  isAttached?: boolean
  /**
   * If `true`, all wrapped button will be disabled
   */
  isDisabled?: boolean
  /**
   * The spacing between the buttons
   * @default '0.5rem'
   * @type SystemProps["marginRight"]
   */
  spacing?: SystemProps["marginRight"]
}

interface ButtonGroupContext extends ThemingProps<"ButtonGroup"> {
  isDisabled?: boolean
}

const [ButtonGroupProvider, useButtonGroup] = createContext<ButtonGroupContext>(
  {
    strict: false,
    name: "ButtonGroupContext",
  },
)

export { useButtonGroup }

export const ButtonGroup = forwardRef<ButtonGroupProps, "div">((props, ref) => {
  const {
    size,
    colorScheme,
    variant,
    className,
    spacing = "0.5rem",
    isAttached,
    isDisabled,
    ...rest
  } = props

  const _className = cx("chakra-button__group", className)

  const context = React.useMemo(
    () => ({ size, colorScheme, variant, isDisabled }),
    [size, colorScheme, variant, isDisabled],
  )

  let groupStyles: SystemStyleObject = {
    display: "inline-flex",
  }

  if (isAttached) {
    groupStyles = {
      ...groupStyles,
      "> *:first-of-type:not(:last-of-type)": { borderEndRadius: 0 },
      "> *:not(:first-of-type):not(:last-of-type)": { borderRadius: 0 },
      "> *:not(:first-of-type):last-of-type": { borderStartRadius: 0 },
    }
  } else {
    groupStyles = {
      ...groupStyles,
      "& > *:not(style) ~ *:not(style)": { marginStart: spacing },
    }
  }

  return (
    <ButtonGroupProvider value={context}>
      <chakra.div
        ref={ref}
        role="group"
        __css={groupStyles}
        className={_className}
        {...rest}
      />
    </ButtonGroupProvider>
  )
})

if (__DEV__) {
  ButtonGroup.displayName = "ButtonGroup"
}
