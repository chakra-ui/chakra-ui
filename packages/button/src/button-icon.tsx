import { chakra, HTMLChakraProps } from "@chakra-ui/system"
import { cx, __DEV__ } from "@chakra-ui/utils"
import * as React from "react"

export const ButtonIcon: React.FC<HTMLChakraProps<"span">> = (props) => {
  const { children, className, ...rest } = props

  const _children = React.isValidElement(children)
    ? React.cloneElement(children, {
        "aria-hidden": true,
        focusable: false,
      })
    : children

  const _className = cx("chakra-button__icon", className)

  return (
    <chakra.span
      display="inline-flex"
      alignSelf="center"
      flexShrink={0}
      {...rest}
      className={_className}
    >
      {_children}
    </chakra.span>
  )
}

if (__DEV__) {
  ButtonIcon.displayName = "ButtonIcon"
}
