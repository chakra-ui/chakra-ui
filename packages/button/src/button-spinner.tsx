import { Spinner } from "@chakra-ui/spinner"
import {
  chakra,
  HTMLChakraProps,
  SystemProps,
  SystemStyleObject,
} from "@chakra-ui/system"
import { cx, __DEV__ } from "@chakra-ui/utils"
import * as React from "react"

interface ButtonSpinnerProps extends HTMLChakraProps<"div"> {
  label?: string
  /**
   * @type SystemProps["margin"]
   */
  spacing?: SystemProps["margin"]
  placement?: "start" | "end"
}

export const ButtonSpinner: React.FC<ButtonSpinnerProps> = (props) => {
  const {
    label,
    placement,
    spacing = "0.5rem",
    children = <Spinner color="currentColor" width="1em" height="1em" />,
    className,
    __css,
    ...rest
  } = props

  const _className = cx("chakra-button__spinner", className)

  const marginProp = placement === "start" ? "marginEnd" : "marginStart"

  const spinnerStyles: SystemStyleObject = React.useMemo(
    () => ({
      display: "flex",
      alignItems: "center",
      position: label ? "relative" : "absolute",
      [marginProp]: label ? spacing : 0,
      fontSize: "1em",
      lineHeight: "normal",
      ...__css,
    }),
    [__css, label, marginProp, spacing],
  )

  return (
    <chakra.div className={_className} {...rest} __css={spinnerStyles}>
      {children}
    </chakra.div>
  )
}
if (__DEV__) {
  ButtonSpinner.displayName = "ButtonSpinner"
}
