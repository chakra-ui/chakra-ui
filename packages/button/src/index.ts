export * from "./button"
export * from "./button-group"
export * from "./icon-button"
export * from "./button-spinner"
