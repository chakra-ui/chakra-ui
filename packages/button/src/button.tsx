import { useMergeRefs } from "@chakra-ui/hooks"
import {
  chakra,
  forwardRef,
  HTMLChakraProps,
  omitThemingProps,
  SystemProps,
  SystemStyleObject,
  ThemingProps,
  useStyleConfig,
} from "@chakra-ui/system"
import { cx, dataAttr, mergeWith, __DEV__ } from "@chakra-ui/utils"
import * as React from "react"
import { useButtonGroup } from "./button-group"
import { ButtonSpinner } from "./button-spinner"
import { ButtonIcon } from "./button-icon"
import { useButtonType } from "./use-button-type"

export interface ButtonOptions {
  /**
   * If `true`, the button will show a spinner.
   */
  isLoading?: boolean
  /**
   * If `true`, the button will be styled in its active state.
   */
  isActive?: boolean
  /**
   * If `true`, the button will be disabled.
   */
  isDisabled?: boolean
  /**
   * The label to show in the button when `isLoading` is true
   * If no text is passed, it only shows the spinner
   */
  loadingText?: string
  /**
   * If `true`, the button will take up the full width of its container.
   */
  isFullWidth?: boolean
  /**
   * The html button type to use.
   */
  type?: "button" | "reset" | "submit"
  /**
   * If added, the button will show an icon before the button's label.
   * @type React.ReactElement
   */
  leftIcon?: React.ReactElement
  /**
   * If added, the button will show an icon after the button's label.
   * @type React.ReactElement
   */
  rightIcon?: React.ReactElement
  /**
   * The space between the button icon and label.
   * @type SystemProps["marginRight"]
   */
  iconSpacing?: SystemProps["marginRight"]
  /**
   * Replace the spinner component when `isLoading` is set to `true`
   * @type React.ReactElement
   */
  spinner?: React.ReactElement
  /**
   * It determines the placement of the spinner when isLoading is true
   * @default "start"
   */
  spinnerPlacement?: "start" | "end"
}

export interface ButtonProps
  extends HTMLChakraProps<"button">,
    ButtonOptions,
    ThemingProps<"Button"> {}

export const Button = forwardRef<ButtonProps, "button">((props, ref) => {
  const group = useButtonGroup()
  const styles = useStyleConfig("Button", { ...group, ...props })

  const {
    isDisabled = group?.isDisabled,
    isLoading,
    isActive,
    isFullWidth,
    children,
    leftIcon,
    rightIcon,
    loadingText,
    iconSpacing = "0.5rem",
    type,
    spinner,
    spinnerPlacement = "start",
    className,
    as,
    ...rest
  } = omitThemingProps(props)

  /**
   * When button is used within ButtonGroup (i.e flushed with sibling buttons),
   * it is important to add a `zIndex` on focus.
   *
   * So let's read the component styles and then add `zIndex` to it.
   */
  const buttonStyles: SystemStyleObject = React.useMemo(() => {
    const _focus = mergeWith({}, styles?.["_focus"] ?? {}, { zIndex: 1 })
    return {
      display: "inline-flex",
      appearance: "none",
      alignItems: "center",
      justifyContent: "center",
      userSelect: "none",
      position: "relative",
      whiteSpace: "nowrap",
      verticalAlign: "middle",
      outline: "none",
      width: isFullWidth ? "100%" : "auto",
      ...styles,
      ...(!!group && { _focus }),
    }
  }, [styles, group, isFullWidth])

  const { ref: _ref, type: defaultType } = useButtonType(as)

  const contentProps = { rightIcon, leftIcon, iconSpacing, children }

  return (
    <chakra.button
      disabled={isDisabled || isLoading}
      ref={useMergeRefs(ref, _ref)}
      as={as}
      type={type ?? defaultType}
      data-active={dataAttr(isActive)}
      data-loading={dataAttr(isLoading)}
      __css={buttonStyles}
      className={cx("chakra-button", className)}
      {...rest}
    >
      {isLoading && spinnerPlacement === "start" && (
        <ButtonSpinner
          className="chakra-button__spinner--start"
          label={loadingText}
          placement="start"
        >
          {spinner}
        </ButtonSpinner>
      )}

      {isLoading ? (
        loadingText || (
          <chakra.span opacity={0}>
            <ButtonContent {...contentProps} />
          </chakra.span>
        )
      ) : (
        <ButtonContent {...contentProps} />
      )}

      {isLoading && spinnerPlacement === "end" && (
        <ButtonSpinner
          className="chakra-button__spinner--end"
          label={loadingText}
          placement="end"
        >
          {spinner}
        </ButtonSpinner>
      )}
    </chakra.button>
  )
})

if (__DEV__) {
  Button.displayName = "Button"
}

type ButtonContentProps = Pick<
  ButtonProps,
  "leftIcon" | "rightIcon" | "children" | "iconSpacing"
>

function ButtonContent(props: ButtonContentProps) {
  const { leftIcon, rightIcon, children, iconSpacing } = props
  return (
    <>
      {leftIcon && <ButtonIcon marginEnd={iconSpacing}>{leftIcon}</ButtonIcon>}
      {children}
      {rightIcon && (
        <ButtonIcon marginStart={iconSpacing}>{rightIcon}</ButtonIcon>
      )}
    </>
  )
}
