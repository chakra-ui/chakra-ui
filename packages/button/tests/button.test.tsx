import * as React from "react"
import { render, testA11y, screen } from "@chakra-ui/test-utils"
import { EmailIcon, ArrowForwardIcon } from "@chakra-ui/icons"
import { Button, ButtonGroup } from "../src"

it("passes a11y test", async () => {
  await testA11y(<Button>test</Button>)
})

test("renders with icons", () => {
  const { getByText } = render(
    <ButtonGroup>
      <Button leftIcon={<EmailIcon />}>Email</Button>
      <Button rightIcon={<ArrowForwardIcon />}>Arrow Forward</Button>
    </ButtonGroup>,
  )
  expect(getByText("Email")).toBeTruthy()
  expect(getByText("Arrow Forward")).toBeTruthy()
})

test("shows spinner if isLoading", () => {
  const { getByTestId } = render(
    <Button data-testid="btn" isLoading>
      Email
    </Button>,
  )
  expect(getByTestId("btn")).toHaveAttribute("data-loading")

  // children text is hidden
  expect(screen.getByText("Email")).not.toBeVisible()

  // "Loading..." visually hidden label shown
  screen.getByText("Loading...")
})

test("shows spinner and loading text if isLoading and loadingText", () => {
  render(
    <Button isLoading loadingText="Submitting">
      Submit
    </Button>,
  )

  // children text is replaced by `loadingText`
  screen.getByText("Submitting")
  expect(screen.queryByText("Submit")).toBeNull()
})

test("has the proper aria attributes", () => {
  const { rerender } = render(<Button>Hello</Button>)

  // button has role="button"
  let button = screen.getByRole("button")
  expect(button).not.toHaveAttribute("aria-disabled")

  // isLoading sets aria-disabled="true"
  rerender(<Button isLoading>Hello</Button>)
  button = screen.getByRole("button")
  expect(button).toHaveAttribute("data-loading", "")

  // isDisabled sets aria-disabled="true"
  rerender(<Button isDisabled>Hello</Button>)
  button = screen.getByRole("button")
  expect(button).toHaveAttribute("disabled", "")
})
