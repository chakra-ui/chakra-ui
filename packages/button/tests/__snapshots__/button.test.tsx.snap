// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: all 250ms;
  transition: all 250ms;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  outline: none;
  width: auto;
  line-height: 1.2;
  border-radius: 0.25rem;
  font-weight: 600;
  height: 2.5rem;
  min-width: 2.5rem;
  font-size: 1rem;
  padding-left: 1rem;
  padding-right: 1rem;
  background: #EDF2F7;
}

.emotion-0:focus,
.emotion-0[data-focus] {
  box-shadow: 0 0 0 3px rgba(66,153,225,0.6);
}

.emotion-0[disabled],
.emotion-0[disabled]:focus,
.emotion-0[disabled]:hover,
.emotion-0[aria-disabled=true],
.emotion-0[aria-disabled=true]:focus,
.emotion-0[aria-disabled=true]:hover,
.emotion-0[data-disabled],
.emotion-0[data-disabled]:focus,
.emotion-0[data-disabled]:hover {
  opacity: 0.4;
  cursor: not-allowed;
  box-shadow: none;
}

.emotion-0:hover,
.emotion-0[data-hover] {
  background: #E2E8F0;
}

.emotion-0:active,
.emotion-0[data-active] {
  background: #CBD5E0;
}

<button
    class="chakra-button emotion-0"
    type="button"
  >
    test
  </button>
</DocumentFragment>
`;

exports[`matches snapshot with icons 1`] = `
<DocumentFragment>
  .emotion-6 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-6 > *:not(style) ~ *:not(style) {
  margin-left: 0.5rem;
}

.emotion-2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: all 250ms;
  transition: all 250ms;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  outline: none;
  width: auto;
  line-height: 1.2;
  border-radius: 0.25rem;
  font-weight: 600;
  height: 2.5rem;
  min-width: 2.5rem;
  font-size: 1rem;
  padding-left: 1rem;
  padding-right: 1rem;
  background: #EDF2F7;
}

.emotion-2:focus,
.emotion-2[data-focus] {
  box-shadow: 0 0 0 3px rgba(66,153,225,0.6);
  z-index: 1;
}

.emotion-2[disabled],
.emotion-2[disabled]:focus,
.emotion-2[disabled]:hover,
.emotion-2[aria-disabled=true],
.emotion-2[aria-disabled=true]:focus,
.emotion-2[aria-disabled=true]:hover,
.emotion-2[data-disabled],
.emotion-2[data-disabled]:focus,
.emotion-2[data-disabled]:hover {
  opacity: 0.4;
  cursor: not-allowed;
  box-shadow: none;
}

.emotion-2:hover,
.emotion-2[data-hover] {
  background: #E2E8F0;
}

.emotion-2:active,
.emotion-2[data-active] {
  background: #CBD5E0;
}

.emotion-1 {
  margin-right: 0.5rem;
}

.emotion-0 {
  width: 1em;
  height: 1em;
  display: inline-block;
  line-height: 1em;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  color: currentColor;
  vertical-align: middle;
}

.emotion-4 {
  margin-left: 0.5rem;
}

<div
    class="chakra-button__group emotion-6"
    role="group"
  >
    <button
      class="chakra-button emotion-2"
      type="button"
    >
      <span
        class="chakra-button__icon emotion-1"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon emotion-0"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <g
            fill="currentColor"
          >
            <path
              d="M11.114,14.556a1.252,1.252,0,0,0,1.768,0L22.568,4.87a.5.5,0,0,0-.281-.849A1.966,1.966,0,0,0,22,4H2a1.966,1.966,0,0,0-.289.021.5.5,0,0,0-.281.849Z"
            />
            <path
              d="M23.888,5.832a.182.182,0,0,0-.2.039l-6.2,6.2a.251.251,0,0,0,0,.354l5.043,5.043a.75.75,0,1,1-1.06,1.061l-5.043-5.043a.25.25,0,0,0-.354,0l-2.129,2.129a2.75,2.75,0,0,1-3.888,0L7.926,13.488a.251.251,0,0,0-.354,0L2.529,18.531a.75.75,0,0,1-1.06-1.061l5.043-5.043a.251.251,0,0,0,0-.354l-6.2-6.2a.18.18,0,0,0-.2-.039A.182.182,0,0,0,0,6V18a2,2,0,0,0,2,2H22a2,2,0,0,0,2-2V6A.181.181,0,0,0,23.888,5.832Z"
            />
          </g>
        </svg>
      </span>
      Email
    </button>
    <button
      class="chakra-button emotion-2"
      type="button"
    >
      Arrow Forward
      <span
        class="chakra-button__icon emotion-4"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon emotion-0"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 4l-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
            fill="currentColor"
          />
        </svg>
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`shows spinner and loading text if isLoading and loadingText 1`] = `
<DocumentFragment>
  @keyframes animation-0 {
  0% {
    -webkit-transform: rotate(0deg);
    -ms-transform: rotate(0deg);
    transform: rotate(0deg);
  }

  100% {
    -webkit-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

.emotion-3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: all 250ms;
  transition: all 250ms;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  outline: none;
  width: auto;
  line-height: 1.2;
  border-radius: 0.25rem;
  font-weight: 600;
  height: 2.5rem;
  min-width: 2.5rem;
  font-size: 1rem;
  padding-left: 1rem;
  padding-right: 1rem;
  background: #EDF2F7;
}

.emotion-3:focus,
.emotion-3[data-focus] {
  box-shadow: 0 0 0 3px rgba(66,153,225,0.6);
}

.emotion-3[disabled],
.emotion-3[disabled]:focus,
.emotion-3[disabled]:hover,
.emotion-3[aria-disabled=true],
.emotion-3[aria-disabled=true]:focus,
.emotion-3[aria-disabled=true]:hover,
.emotion-3[data-disabled],
.emotion-3[data-disabled]:focus,
.emotion-3[data-disabled]:hover {
  opacity: 0.4;
  cursor: not-allowed;
  box-shadow: none;
}

.emotion-3:hover,
.emotion-3[data-hover] {
  background: #E2E8F0;
}

.emotion-3:active,
.emotion-3[data-active] {
  background: #CBD5E0;
}

.emotion-1 {
  display: inline-block;
  border-color: currentColor;
  border-style: solid;
  border-radius: 99999px;
  border-width: 2px;
  border-bottom-color: transparent;
  border-left-color: transparent;
  color: currentColor;
  -webkit-animation: animation-0 0.45s linear infinite;
  animation: animation-0 0.45s linear infinite;
  width: 1em;
  height: 1em;
}

.emotion-0 {
  border: 0px;
  -webkit-clip: rect(0px,0px,0px,0px);
  clip: rect(0px,0px,0px,0px);
  height: 1px;
  width: 1px;
  margin: -1px;
  padding: 0px;
  overflow: hidden;
  white-space: nowrap;
  position: absolute;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  margin-right: 0.5rem;
  font-size: 1em;
  line-height: normal;
}

<button
    class="chakra-button emotion-3"
    data-loading=""
    disabled=""
    type="button"
  >
    <div
      class="chakra-button__spinner emotion-2"
    >
      <div
        class="chakra-spinner emotion-1"
      >
        <span
          class="emotion-0"
        >
          Loading...
        </span>
      </div>
    </div>
    Submitting
  </button>
</DocumentFragment>
`;

exports[`shows spinner if isLoading 1`] = `
<DocumentFragment>
  @keyframes animation-0 {
  0% {
    -webkit-transform: rotate(0deg);
    -ms-transform: rotate(0deg);
    transform: rotate(0deg);
  }

  100% {
    -webkit-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

.emotion-4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: all 250ms;
  transition: all 250ms;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  outline: none;
  width: auto;
  line-height: 1.2;
  border-radius: 0.25rem;
  font-weight: 600;
  height: 2.5rem;
  min-width: 2.5rem;
  font-size: 1rem;
  padding-left: 1rem;
  padding-right: 1rem;
  background: #EDF2F7;
}

.emotion-4:focus,
.emotion-4[data-focus] {
  box-shadow: 0 0 0 3px rgba(66,153,225,0.6);
}

.emotion-4[disabled],
.emotion-4[disabled]:focus,
.emotion-4[disabled]:hover,
.emotion-4[aria-disabled=true],
.emotion-4[aria-disabled=true]:focus,
.emotion-4[aria-disabled=true]:hover,
.emotion-4[data-disabled],
.emotion-4[data-disabled]:focus,
.emotion-4[data-disabled]:hover {
  opacity: 0.4;
  cursor: not-allowed;
  box-shadow: none;
}

.emotion-4:hover,
.emotion-4[data-hover] {
  background: #E2E8F0;
}

.emotion-4:active,
.emotion-4[data-active] {
  background: #CBD5E0;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: absolute;
  margin-right: 0;
  font-size: 1em;
  line-height: normal;
}

.emotion-1 {
  display: inline-block;
  border-color: currentColor;
  border-style: solid;
  border-radius: 99999px;
  border-width: 2px;
  border-bottom-color: transparent;
  border-left-color: transparent;
  color: currentColor;
  -webkit-animation: animation-0 0.45s linear infinite;
  animation: animation-0 0.45s linear infinite;
  width: 1em;
  height: 1em;
}

.emotion-0 {
  border: 0px;
  -webkit-clip: rect(0px,0px,0px,0px);
  clip: rect(0px,0px,0px,0px);
  height: 1px;
  width: 1px;
  margin: -1px;
  padding: 0px;
  overflow: hidden;
  white-space: nowrap;
  position: absolute;
}

.emotion-3 {
  opacity: 0;
}

<button
    class="chakra-button emotion-4"
    data-loading=""
    disabled=""
    type="button"
  >
    <div
      class="chakra-button__spinner emotion-2"
    >
      <div
        class="chakra-spinner emotion-1"
      >
        <span
          class="emotion-0"
        >
          Loading...
        </span>
      </div>
    </div>
    <span
      class="emotion-3"
    >
      Email
    </span>
  </button>
</DocumentFragment>
`;
